subsys/net/buf.c

```diff
@@ -502,6 +501,7 @@ void net_buf_unref(struct net_buf *buf)
 		if (pool->destroy) {
 			pool->destroy(buf);
 		} else {
+			buf->data = NULL;
```

Comment:
Hmm, why not place the setting after this if() block? Then the PR would leave the net_buf in the same state as before. Probably this does not matter but now buf->data contains address even after it is destroyed.

Reply:
Doesn't doing so open up a race? When can the destroyed pool item be re-allocated?

Reply:
An alternative would be to move this to `net_buf_destroy()` itself, where for the current uses in-tree the outcome would be unchanged (all custom destroy callbacks do call `net_buf_destroy()`, but I am not sure if it would be proper place to do so.

Reply:
Yes, indeed my proposal would be racy so lets not do that.

Reply:
You're mixing `buf->data` and `buf->__buf`.

`buf->__buf` is the actual storage buffer, and this pointer stays unchanged over the lifetime of the net_buf. `buf->data` otoh can point anywhere between `buf->__buf` and `buf->__buf + buf->size` (depending on what kind of encoding/decoding operations have been performed on the buffer).